app/models/concerns/orchestration.rb

```diff
@@ -142,20 +145,26 @@ def execute(opts = {})
   end
 
   # we keep the before update host object in order to compare changes
-  def setup_clone
+  def setup_clone(&block)
     return if new_record?
-    @old = dup
-    for key in (changed_attributes.keys - ["updated_at"])
-      @old.send "#{key}=", changed_attributes[key]
-      # At this point the old cached bindings may still be present so we force an AR association reload
-      # This logic may not work or be required if we switch to Rails 3
-      if (match = key.match(/\A(.*)_id\Z/))
-        name = match[1].to_sym
-        next if name == :owner # This does not work for the owner association even from the console
-        self.send(name, true) if (send(name) and send(name).id != @attributes[key])
-        old.send(name, true)  if (old.send(name) and old.send(name).id != old.attributes[key])
-      end
+    @old = setup_object_clone(self, &block)
+      # NOTE for reviewers - I tested in rails 2 and there was an issue like this
+      #   host.architecture_id      # => 1
+      #   host.architecture_id = 2
+      #   host.architecture         # => <Architecture id: 1...
+      # but in Rails 3 it gives us  # => <Architecture id: 2...
+      # so we don't need it anymore
+      # TODO remove this comment after review and before merge
+  end
+
+  def setup_object_clone(object)
+    clone = object.dup
```

Comment:
@eLobato do you refer to `for key in ...` code below? I think deep_clone does not help much, IIRC we have to change attributes after we yield a block but we need to run the block on already existing clone